### ISSUE TITLE
Improve the flow of host permission granting

### ISSUE THREAD

**Issue by DNin01** [ScratchAddons/ScratchAddons#7531]
#### 1. To be nicer

I believe since MV3, Firefox doesn't grant any of the host permissions that we declare as required on install, so we have to grant them during runtime. That means we will have to ask the user to give us those permissions.

We haven't done this sort of thing before, so currently we just re-use our existing permissions page, which is worded like this:

> ## Please enable site access
>
> Scratch Addons needs site access permission to work.
>
> ### `Enable`
> <sub>Click the button to enable.</sub>
>
> <img src=https://raw.githubusercontent.com/ScratchAddons/ScratchAddons/master/images/screenshots/permissions-light.png width=400>
>
> <sub>Click "Allow". Scratch Addons will reload and the addons will start working correctly again</sub>

#### 2. A better way to ask for permissions
In the future, we may want to add more host permissions. However, declaring permissions as required disables an extension until those permissions are granted[^1]. I'm not sure if that's the case for host permissions. To get around this, we would need to declare those permissions as optional so that they don't block the extension if they aren't granted.

I, of course, knowing what was going on, would grant Scratch Addons any new permissions it asks for if I can think of why it might need them (such as access to its own website), but some other people who haven't seen such a prompt before and maybe aren't thinking clearly in the moment may end up denying the permission and never using the extension again.

With the new scratch.org website on the horizon[^2], we could need to do this in the near future.

[^1]:
     > When a new permission that [triggers a warning](https://developer.chrome.com/docs/extensions/reference/permissions-list) is added, the extension will be disabled until the user accepts the new permission. See [Updating permissions](https://developer.chrome.com/docs/extensions/develop/concepts/permission-warnings#update_permissions) to learn how to test this behavior.

     https://developer.chrome.com/docs/extensions/develop/concepts/permission-warnings

[^2]:
     **Heads up! Learn about scratch.org and new link redirects**: https://scratch.mit.edu/discuss/topic/765450/

     Or see #6508

#### A better permission prompt

I was thinking of a page like this:

![](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/e9932903-b93e-4e72-96e7-5482845d22aa)

This page (which might remind you of cookie settings modals) makes it clear why each permission needs to be granted and lists each website that we want access to and why. There are even switches to the right of each listed website that the user can turn off if they don't want to give us access to particular sites (not that lots of users would do that, but just in case, and to lighten the pressure).

This page could be shown during onboarding as part of a sequence of steps. It could be the first thing that appears after clicking a "Get Started" button.

#### Make more permissions optional

Currently, every host permission is a soft requirement. Opening the Scratch Addons popup or setting page without every single one checked will display the permissions page shown earlier. We can do that, but it's also misleading and unnecessary - the extension still works with all but scratch.mit.edu allowed, but some background processes and popup features don't work. I believe users should be allowed to disable a permission that they know isn't needed without being nagged.

In the future, we could work on granting these permissions as part of enabling addons that require them. This might allow us to not even show them as options on that permissions page.

#### Add more permissions

This would also be a great opportunity to declare some more host permissions. If we can ask for permissions ourselves, then, like I said before, we wouldn't need the browser to tell the user to grant us those permissions before Scratch Addons can be used again.

### Additional context

With this implementation, we may even decide to make all permissions optional and show them to the user during onboarding, even on browsers that grant us required host permissions during installation. That means Scratch Addons, whether used on Chrome or Firefox, would have the same flow of allowing site access.

We never open pages that aren't associated with a user interaction unless we have to, and we wouldn't want to do this just to get the user to give us permissions we haven't asked for before. Bad UX around permissions could lead to uninstalls.

We could also redesign the page that appears when host permissions are revoked:

![](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/4343859b-d0f3-487e-9aee-51d88f5a5f69)

Clicking "Enable" would ask for all permissions, which the user would be prompted for by the browser. Clicking "More options" would go to the site access options page shown earlier.

---

<sub>Font used in mock-ups: Sora. Copyright 2019 The Sora Project Authors (https://github.com/sora-xor/sora-font)</sub>

**Comment by Samq64** [comment]
Alternatively the more options button could just open the browser's permission page.

**Comment by WorldLanguages** [comment]
> With this implementation, we may even decide to make all permissions optional and show them to the user during onboarding, even on browsers that grant us required host permissions during installation. That means Scratch Addons, whether used on Chrome or Firefox, would have the same flow of allowing site access.

Making all host permissions optional is planned by the Chrome extensions team, and we won't have a choice. The final design is yet to be seen, but I'm pretty sure Chrome wants to go away from granting host permissions when installing.

**Comment by WorldLanguages** [comment]
I don't think it will ever be necessary to ask for a scratchaddons.com host permission (we can always use a Scratch iframe if needed and configure CORS properly in our own website) but if we want, we could assume it is for the purpose of this UI and avoid any requests to it if it's disabled. Probably not a good idea?

Having separate permissions for scratch.mit.edu, api and clouddata makes very little sense as well. Remember that the scratch.mit.edu website can already access the other two, so the first permission implies the two others. If users want to disable all website addons but continue to use Scratch Messaging, we can provide something more intuitive instead, such as an extension pause button or disable toggle (there's more discussion in other issues).

I can only see the UI making sense if we add toggles for more sites, such as running editor addons on ScratchLab, CSFirst, etc.

**Comment by DNin01** [comment]
> I don't think it will ever be necessary to ask for a scratchaddons.com host permission (we can always use a Scratch iframe if needed and configure CORS properly in our own website) but if we want, we could assume it is for the purpose of this UI and avoid any requests to it if it's disabled. Probably not a good idea?

Then the only benefit to requesting permission to scratchaddons.com would be that the user would understand why we are able to access it. One would expect that we need that permission at first, right? But actually, we can just access it through a content script on a Scratch page we have access to.

On the other hand, in general, it makes sense to request as few permissions as possible, so not asking for this permission would be one less impediment to getting started with Scratch Addons.

> Having separate permissions for scratch.mit.edu, api and clouddata makes very little sense as well. Remember that the scratch.mit.edu website can already access the other two, so the first permission implies the two others. If users want to disable all website addons but continue to use Scratch Messaging, we can provide something more intuitive instead, such as an extension pause button or disable toggle (there's more discussion in other issues).

For users that don't know everything about the permissions system, that seems like a reasonable assumption. I don't think anyone would be surprised that we could access those Scratch endpoints through a single host permission to scratch.mit.edu.

**Comment by WorldLanguages** [comment]
To clarify, requesting the api and clouddata permissions directly is handy, as we can reach those subdomains from the background context with a normal fetch. But in theory, we could also access those two by making the requests from a scratch.mit.edu tab or background iframe.

**Comment by WorldLanguages** [comment]
> Then the only benefit to requesting permission to scratchaddons.com would be that the user would understand why we are able to access it.

I'm not the average user, but of course extensions can access public information where CORS headers are correctly configured. Some addons already do this with ScratchDB and Ocular.

**Comment by DNin01** [comment]
> > Then the only benefit to requesting permission to scratchaddons.com would be that the user would understand why we are able to access it.
> 
> I'm not the average user, but of course extensions can access public information where CORS headers are correctly configured. Some addons already do this with ScratchDB and Ocular.

Can that be done even from within extension pages without host permissions? I've never actually looked to see if an extension can make any fetch it wants or only fetches to sites that it has been granted a host permission to.

Either way, I'd bet if any requests were blocked by CORS, a host permission would be required.

**Comment by DNin01** [comment]
So, extension pages _can_ fetch anything, but I tried a bunch of Scratch URLs and the only one I was able to fetch from was an embeddable project page; all the others were blocked due to CORS. But other websites can access them... It's probably because extension pages' scheme is `chrome-extension:` but the requests we're trying to make are to `https:` pages.

**Comment by WorldLanguages** [comment]
> But other websites can access them

No...?
Also, you want to distinguish between 1) making a request, 2) making a request with credentials/cookies, and 3) being able to read the response.
